components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java

```diff
@@ -0,0 +1,100 @@
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+package org.wso2.carbon.configuration;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.carbon.configuration.provider.ConfigProvider;
+import org.wso2.carbon.configuration.provider.ConfigProviderDataHolder;
+import org.wso2.carbon.configuration.provider.ConfigProviderImpl;
+import org.wso2.carbon.configuration.reader.ConfigFileReader;
+import org.wso2.carbon.configuration.reader.YAMLBasedConfigFileReader;
+import org.wso2.carbon.securevault.SecureVault;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * This service component is responsible for registering ConfigProvider OSGi service.
+ *
+ * @since 5.2.0
+ */
+@Component(
+        name = "org.wso2.carbon.configuration.ConfigProviderComponent",
+        immediate = true
+)
+public class ConfigProviderComponent {
+    private static final Logger logger = LoggerFactory.getLogger(ConfigProviderComponent.class);
+
+    @Activate
+    protected void activate(BundleContext bundleContext) {
+        ConfigProviderDataHolder.getInstance().setBundleContext(bundleContext);
+        initializeConfigProvider();
+        logger.debug("Activating Carbon Configuration Component");
+    }
+
+    @Deactivate
+    protected void deactivate(BundleContext bundleContext) {
+        ConfigProviderDataHolder.getInstance().setBundleContext(null);
+        logger.debug("Stopping ConfigProviderComponent");
+    }
+
+    @Reference(
+            name = "org.wso2.carbon.securevault",
+            service = SecureVault.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unRegisterSecureVault"
+    )
+    protected void registerSecureVault(SecureVault secureVault) {
+        ConfigProviderDataHolder.getInstance().setSecureVault(secureVault);
+        initializeConfigProvider();
+    }
+
+    protected void unRegisterSecureVault(SecureVault secureVault) {
+        ConfigProviderDataHolder.getInstance().setSecureVault(null);
+    }
+
+    /**
+     * Initialise carbon config provider.
+     */
+    private void initializeConfigProvider() {
+        if (!ConfigProviderDataHolder.getInstance().getSecureVault().isPresent() ||
+                !ConfigProviderDataHolder.getInstance().getBundleContext().isPresent()) {
+            logger.debug("Waiting for config provider dependencies");
+            return;
+        }
```

Comment:
Yes, Carbon config is dependant on Secure Vault. Therefore we only need to register the configuration service once the secure vault service is registered as an OSGi service